repo/grpc_repository_client.go

```diff
@@ -0,0 +1,532 @@
+	"github.com/kopia/kopia/repo/hashing"
+	"github.com/kopia/kopia/repo/manifest"
+	"github.com/kopia/kopia/repo/object"
+)
+
+// MaxGRPCMessageSize is the maximum size of a message sent or received over GRPC API when talking to
+// Kopia repository server. This is bigger than the size of any possible content, which is
+// defined by supported splitters.
+const MaxGRPCMessageSize = 20 << 20
+
+func errNoSessionResponse() error {
+	return errors.New("did not receive response from the server")
+}
+
+// grpcRepositoryClient is an implementation of Repository that connects to an instance of
+// GPRC API server hosted by `kopia server`.
+type grpcRepositoryClient struct {
+	connRefCount *int32
+	conn         *grpc.ClientConn
+	sess         apipb.KopiaRepository_SessionClient
+
+	h            hashing.HashFunc
+	objectFormat object.Format
+	cliOpts      ClientOptions
+	omgr         *object.Manager
+
+	sendMutex sync.Mutex
+
+	activeRequestsMutex sync.Mutex
+	nextRequestID       int64
+	activeRequests      map[int64]chan *apipb.SessionResponse
+}
+
+// readLoop runs in a goroutine and consumes all messages in session and forwards them to appropriate channels.
+func (r *grpcRepositoryClient) readLoop() {
+	msg, err := r.sess.Recv()
+
+	for ; err == nil; msg, err = r.sess.Recv() {
+		r.activeRequestsMutex.Lock()
+		ch := r.activeRequests[msg.RequestId]
+		delete(r.activeRequests, msg.RequestId)
+
+		r.activeRequestsMutex.Unlock()
+
+		ch <- msg
+		close(ch)
+	}
+
+	// when a read loop error occurs, close all pending client channels with an artificial error.
+	r.activeRequestsMutex.Lock()
+	defer r.activeRequestsMutex.Unlock()
+
+	errResponse := &apipb.SessionResponse{
+		Response: &apipb.SessionResponse_Error{
+			Error: &apipb.ErrorResponse{
+				Code:    apipb.ErrorResponse_UNKNOWN_ERROR,
+				Message: err.Error(),
+			},
+		},
+	}
+
+	for id, ch := range r.activeRequests {
+		delete(r.activeRequests, id)
+
+		ch <- errResponse
+		close(ch)
+	}
+}
+
+// sendRequest sends the provided request to the server and returns a channel on which the
+// caller can receive session response(s).
+func (r *grpcRepositoryClient) sendRequest(req *apipb.SessionRequest) chan *apipb.SessionResponse {
+	// allocate request ID and create channel to which we're forwarding the responses.
+	r.activeRequestsMutex.Lock()
+	rid := r.nextRequestID
+	r.nextRequestID++
+
+	ch := make(chan *apipb.SessionResponse, 1)
+
+	r.activeRequests[rid] = ch
+	r.activeRequestsMutex.Unlock()
+
+	req.RequestId = rid
+
+	// sends to GRPC stream must be single-threaded.
+	r.sendMutex.Lock()
+	defer r.sendMutex.Unlock()
+
+	// try sending the request and if unable to do so, stuff an error response to the channel
+	// to simplify client code.
+	if err := r.sess.Send(req); err != nil {
+		ch <- &apipb.SessionResponse{
+			Response: &apipb.SessionResponse_Error{
+				Error: &apipb.ErrorResponse{
+					Code:    apipb.ErrorResponse_CLIENT_ERROR,
+					Message: err.Error(),
+				},
+			},
+		}
```

Comment:
Should the channel be closed in this case, and also removed from the active requests?

Reply:
added close()

Reply:
But then, that would result in a double close of the channel, both here and at the bottom of `readLoop`, which will then panic.